### DISCUSSION TITLE
Concern: excessive bureaucracy may be harming the long-term health of the project

### DISCUSSION THREAD

**Discussion by tecosaur** [gethomepage/homepage#4204] [General]
Hello!

I've recently become aware of this project. While setting up the search widget, I noticed the absence of Kagi, a relatively popular non-Google search provider. Since adding it took just a minor tweak, I spun up small PR.

Through this, I've discovered that the 10-upvote policy seems to apply indiscriminately to features, and enhancements to existing features, of all sizes. This is quite a surprise to me, and out of step with the approach taken by all other open-source projects I've been involved with.

As a maintainer of several project myself, I deeply sympathise with the need to consider the maintenance burden of additions. The burden varies massively depending on the particular contribution, separately from how much general interest an idea attracts. In these regards, the "upvote policy" seems like a poor crutch to me, as it is ultimately maintainers who are best placed to judge whether a contribution is worth it, not the drive-by public. I see there's some mention about needing to apply the threshold bluntly to not be "inconsistent", which confuses me as one of the defining characteristics of most projects is *maintainer's taste*. To me, "inconsistency" just sounds like exercising this taste. Some of the best projects I know can credit their quality and success to the maintainer's excellent taste in shaping the project, contributions, and developing contributors. Applying the same upvote threshold to all non-bugfix PRs seems senseless and ultimately counter-productive.

Unfortunately, looking at the past few weeks of PRs, there are plenty of disheartening examples of what look like small, worthwhile contributions being pushed away because they haven't been linked to a 10+ upload threshold:
- https://github.com/gethomepage/homepage/pull/4198 (my PR)
- https://github.com/gethomepage/homepage/pull/4196
- https://github.com/gethomepage/homepage/pull/4102
- https://github.com/gethomepage/homepage/pull/4094
- https://github.com/gethomepage/homepage/pull/4060

This gets onto what I fear the largest impact of this policy is: driving away potential repeat contributors. I consider myself a rather tenacious contributor (hence me going to the effort to make this issue), and I felt quite put-off by the "sorry, computer says no"-style response. Worrying that other people will have had the same, if not a stronger reaction, I've looked at the contribution history in this repo, and am concerned by what I see.

Over the this project two-year lifetime, there have been 352 different contributors, according to GitHub. Since the policy was put into place late last year, there have been just 20 (excluding maintainers and bots). In the past six months, there were only five.

![image](https://github.com/user-attachments/assets/931ce2de-31ce-4d5a-8d9a-09d307783cf3)

Perhaps this project has reached a stage of maturity where external contributions are of little interest, or not needed. In this case, the current policy, along with the decision to lock every single issue a month after closing, seem highly effective at driving contributors away. I appreciate the need to avoid being flooded with features that you're not able or willing to maintain, but the adjustments made look like a gross over-reaction to me.

I hope that I've missed something here, and that this isn't the intent, but it doesn't look good. I hope you'll reconsider whether this policy is in the best interests of the project.

**Comment by shamoon** [comment]
I don’t really appreciate this. I think the project is doing fine and maintains a sustainable pace. You have known about it for about five minutes and drawn a conclusion that could just as easily be explained by the fact that the project isn’t evolving as much now that it’s matured. The other maintainer and I have explicitly stated we don’t want certain things here like user management etc. so yes, PRs are smaller, mostly widgets and development has chilled out.

A lot of people assume their PR is “no maintenance” and that is often not the case. Moreover, when we did not apply the up-vote rule uniformly we received flak that we were being inconsistent and capricious, so here we are.

Locking threads is to prevent people commenting on things that are closed and not getting a response, nothing prevents opening a new one about the same issue. We read and respond to the vast majority of threads, most of which are basic support kinda stuff (and when not, we investigate and address as needed).

Finally, if a Kagi option was really that popular shouldn’t it be pretty easy to get votes? At least it would be some indication that the PR isn’t just for you and nobody else (which is the point here). The same applies more generally.

I make no special claim to this project other than spending a lot of time working on it, which I am happy to back away from if that’s what’s needed. In my experience, most people underestimate what it means to sustain a project over a long period of time.

**Reply by tecosaur** [reply]
I don't mean to denigrate the great work you clearly have been doing here at all. I'm just concerned that the particular approach taken strikes a poor balance of maintainer needs and a reasonable hurdle for eager contributers.

> The other maintainer and I have explicitly stated we don’t want certain things here like user management etc.

Right. I'd just respectfully suggest that when you get PRs that are outside the scope of the project, you just… say that. I'd hope that other PRs would be primarily evaluated by looking at the details of them and using your skill/knowledge of the project.

The PRs I linked are all quite small in scope, and look in line with the work already in the project. Like you say, I've only known this project for about five minutes, but it would be good to hear why you are a maintainer don't think the PRs are a good idea. Speaking as a contributor, having work procedurally rejected for reasons separate to the work itself is disheartening.

I see that you've just updated the guidelines to no longer mention a 10 upvote requirement,  hope and suspect this will be beneficial to the project 🙂

**Comment by shamoon** [comment]
I have no problem relaxing the requirement for non-bugfix PRs, I'll update the docs. We'll still maintain it for service widgets.

Thanks for your input.

**Comment by shamoon** [comment]
Just a final thought as to why the stats above are such bs, the majority of the commits from the beginning of this project were from the weblate translation bot which is obvious when looking at the full picture below.

I appreciate you had good intentions but I think you twisted things to prove a point about something you know very little.

![Contributors to gethomepagehomepage jpeg](https://github.com/user-attachments/assets/c91786a9-fa38-4e47-b148-e4bd9bc1ed2a)
